{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "high_tcr=pd.read_csv(\"1138_highTCR.csv\")\n",
    "mid_tcr=pd.read_csv(\"1182_midTCR.csv\")\n",
    "low_tcr=pd.read_csv(\"1260_lowTCR.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Function make_intial creates a dataframe from a sample ID. If using different filename, this must be changed within\n",
    "the function. Standard file naming: each file should start with the sample and the rest of the filename should be the \n",
    "same between other samples. \n",
    "\n",
    "The intitial dataframe has the counts at each bp position of every gene that was in the original bedfile. These\n",
    "counts are raw and are not normalized to the number of reads or TTs in the region.\n",
    "'''\n",
    "def make_initial(sample):\n",
    "    plus_df=pd.read_table(\"primary_TSS_bedfiles/\"+sample+\"_plus_TSS_xrseq_2.bed\",header=None)\n",
    "    plus_df.columns=['chr','start','end','type','gene','strand','position','plus_count']\n",
    "    minus_df=pd.read_table(\"primary_TSS_bedfiles/\"+sample+\"_minus_TSS_xrseq_2.bed\",header=None)\n",
    "    minus_df.columns=['chr','start','end','type','gene','strand','position','minus_count']\n",
    "    merge_df=pd.merge(plus_df,minus_df)\n",
    "    plusgenes_df=merge_df[merge_df['strand']=='+']\n",
    "    plusgenes_df.columns=['chr','start','end','type','gene','strand','position','NTS','TS']\n",
    "    minusgenes_df=merge_df[merge_df['strand']=='-']\n",
    "    minusgenes_df.columns=['chr','start','end','type','gene','strand','position','TS','NTS']\n",
    "    joined_df=pd.concat([plusgenes_df,minusgenes_df],sort=True)\n",
    "    joined_df=joined_df.sort_index()\n",
    "    joined_df=joined_df.loc[:,['gene','strand','position','TS','NTS']]\n",
    "    \n",
    "    return joined_df\n",
    "\n",
    "'''\n",
    "Function norm_TT takes a dataframe (from make initial function) and normalizes the TS and NTS counts based on the \n",
    "number of TTs on each strand whithin a specified region.\n",
    "\n",
    "The TT_content.py script and bedfile with genomic regions used to make the sample files should be used in order to have \n",
    "the correct normalization.\n",
    "'''\n",
    "def norm_TT(joined_df,TT_file,bedfile):\n",
    "    TT_counts=pd.read_csv(TT_file,sep='\\t',index_col=0)\n",
    "    TT_counts.columns=['TT_TS','TT_NTS']\n",
    "    gene_names=pd.read_csv(bedfile,sep='\\t',header=None,usecols=[4],names=['gene'])\n",
    "    TT_counts['gene']=gene_names['gene']\n",
    "    \n",
    "    joined_df=joined_df.merge(TT_counts,on='gene')\n",
    "    joined_df['norm_TS']=joined_df['TS']/joined_df['TT_TS']\n",
    "    joined_df['norm_NTS']=joined_df['NTS']/joined_df['TT_NTS']\n",
    "    joined_df=joined_df.loc[:,['gene','strand','position','norm_TS','norm_NTS']]\n",
    "    \n",
    "    return joined_df\n",
    "'''\n",
    "Function make_vector will take the average of all counts between genes.\n",
    "It also flips the reverse genes so that they will be in the same orientation as the forward genes.\n",
    "'''\n",
    "def make_vector(joined_df,s):\n",
    "    forward_genes=joined_df[joined_df['strand']=='+']\n",
    "    forward_genes=forward_genes.groupby(['position']).mean()\n",
    "    forward_lis=forward_genes['norm_'+s].tolist()\n",
    "    reverse_genes=joined_df[joined_df['strand']=='-']\n",
    "    reverse_genes=reverse_genes.groupby(['position']).mean()\n",
    "    reverse_lis=reverse_genes['norm_'+s].tolist()\n",
    "    reverse_lis=reverse_lis[::-1]\n",
    "    forward_vector=np.array(forward_lis)\n",
    "    reverse_vector=np.array(reverse_lis)\n",
    "    sum_vector= forward_vector + reverse_vector\n",
    "\n",
    "    return sum_vector\n",
    "'''\n",
    "Function make_final makes the final DF by converting the vectors for TS and NTS into \n",
    "dataframe columns. It then uses a window sizes to compute the rolling average of each\n",
    "column.\n",
    "'''\n",
    "def make_final(joined_df,win_size):\n",
    "    sum_vector_TS=make_vector(joined_df,'TS')\n",
    "    sum_vector_NTS=make_vector(joined_df,'NTS')\n",
    "    sum_df=pd.DataFrame({'TS':sum_vector_TS, 'NTS':sum_vector_NTS})\n",
    "    final_df=sum_df.rolling(window=win_size,min_periods=1).mean()\n",
    "    \n",
    "    return final_df\n",
    "    \n",
    "\n",
    "'''\n",
    "Function make_df takes in a sample name, a TT_file, Bedfile and optional filter_df\n",
    "Bedfile should have been used to make the TT_counts file with the TT_content.py script. TT_counts is just the number \n",
    "of TTs on each strand for the specified regions in the bedfile. These two files are important for normalizing by the \n",
    "number of TTs in each strand.\n",
    "\n",
    "The optional filter file is a csv of genes that should be used to filter the dataframe.(csv should just be one\n",
    "column of gene names and the column name MUST be gene)\n",
    "'''\n",
    "\n",
    "def make_df(sample,TT_file,bedfile,filter_genes=None):\n",
    "    \n",
    "    joined_df=make_initial(sample)\n",
    "    \n",
    "    if filter_genes is not None:\n",
    "        filter_df=pd.read_csv(filter_genes)\n",
    "        joined_df=joined_df.merge(filter_df,on='gene')\n",
    "    \n",
    "    \n",
    "    joined_df=norm_TT(joined_df,TT_file,bedfile)\n",
    "    \n",
    "    final_df=make_final(joined_df,125)\n",
    "    \n",
    "    return final_df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sample_dict="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "TT_file='TTcount.txt'\n",
    "bedfile='primary_upstreamsense_TSS_2'\n",
    "high_tcr=\"1138_highTCR.csv\"\n",
    "\n",
    "\n",
    "\n",
    "wt120_df=make_df('bm01',TT_file,bedfile,high_tcr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TS</th>\n",
       "      <th>NTS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.662186</td>\n",
       "      <td>0.771330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.670423</td>\n",
       "      <td>0.747816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.664781</td>\n",
       "      <td>0.770311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.651599</td>\n",
       "      <td>0.786940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.648017</td>\n",
       "      <td>0.794297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.645058</td>\n",
       "      <td>0.817144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.648140</td>\n",
       "      <td>0.834525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.655704</td>\n",
       "      <td>0.839794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.663299</td>\n",
       "      <td>0.842595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.662293</td>\n",
       "      <td>0.848476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.660714</td>\n",
       "      <td>0.854608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.664256</td>\n",
       "      <td>0.860717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.669551</td>\n",
       "      <td>0.868098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.678716</td>\n",
       "      <td>0.866565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.684865</td>\n",
       "      <td>0.863260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.691769</td>\n",
       "      <td>0.858660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.696581</td>\n",
       "      <td>0.853784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.701039</td>\n",
       "      <td>0.848877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.706256</td>\n",
       "      <td>0.837904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.713499</td>\n",
       "      <td>0.827235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.716584</td>\n",
       "      <td>0.816898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.718426</td>\n",
       "      <td>0.806442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.738118</td>\n",
       "      <td>0.842591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.758271</td>\n",
       "      <td>0.876004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.779711</td>\n",
       "      <td>0.907349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.797201</td>\n",
       "      <td>0.933334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.811674</td>\n",
       "      <td>0.957355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.825369</td>\n",
       "      <td>0.986242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0.851571</td>\n",
       "      <td>1.013064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0.875812</td>\n",
       "      <td>1.038352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>970</th>\n",
       "      <td>1.362286</td>\n",
       "      <td>0.487075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>1.364332</td>\n",
       "      <td>0.485861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>972</th>\n",
       "      <td>1.366952</td>\n",
       "      <td>0.484254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>1.370742</td>\n",
       "      <td>0.482999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>1.374617</td>\n",
       "      <td>0.482199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>975</th>\n",
       "      <td>1.379062</td>\n",
       "      <td>0.481410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>1.383927</td>\n",
       "      <td>0.480518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>977</th>\n",
       "      <td>1.386315</td>\n",
       "      <td>0.479590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978</th>\n",
       "      <td>1.390629</td>\n",
       "      <td>0.478590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>979</th>\n",
       "      <td>1.395271</td>\n",
       "      <td>0.478092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>980</th>\n",
       "      <td>1.399911</td>\n",
       "      <td>0.477632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>981</th>\n",
       "      <td>1.403010</td>\n",
       "      <td>0.477391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982</th>\n",
       "      <td>1.407117</td>\n",
       "      <td>0.477618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>1.410680</td>\n",
       "      <td>0.477404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>1.414096</td>\n",
       "      <td>0.477067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985</th>\n",
       "      <td>1.416664</td>\n",
       "      <td>0.477249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>1.418140</td>\n",
       "      <td>0.477510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>1.419588</td>\n",
       "      <td>0.478258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>1.420723</td>\n",
       "      <td>0.478977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>1.421980</td>\n",
       "      <td>0.479671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>1.421662</td>\n",
       "      <td>0.480670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>1.420208</td>\n",
       "      <td>0.481684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>1.417377</td>\n",
       "      <td>0.482586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>1.416061</td>\n",
       "      <td>0.483551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>1.414663</td>\n",
       "      <td>0.484430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>1.414273</td>\n",
       "      <td>0.485187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>1.414419</td>\n",
       "      <td>0.485839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>1.414960</td>\n",
       "      <td>0.486591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>1.414727</td>\n",
       "      <td>0.487044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>1.413222</td>\n",
       "      <td>0.489630</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TS       NTS\n",
       "0    0.662186  0.771330\n",
       "1    0.670423  0.747816\n",
       "2    0.664781  0.770311\n",
       "3    0.651599  0.786940\n",
       "4    0.648017  0.794297\n",
       "5    0.645058  0.817144\n",
       "6    0.648140  0.834525\n",
       "7    0.655704  0.839794\n",
       "8    0.663299  0.842595\n",
       "9    0.662293  0.848476\n",
       "10   0.660714  0.854608\n",
       "11   0.664256  0.860717\n",
       "12   0.669551  0.868098\n",
       "13   0.678716  0.866565\n",
       "14   0.684865  0.863260\n",
       "15   0.691769  0.858660\n",
       "16   0.696581  0.853784\n",
       "17   0.701039  0.848877\n",
       "18   0.706256  0.837904\n",
       "19   0.713499  0.827235\n",
       "20   0.716584  0.816898\n",
       "21   0.718426  0.806442\n",
       "22   0.738118  0.842591\n",
       "23   0.758271  0.876004\n",
       "24   0.779711  0.907349\n",
       "25   0.797201  0.933334\n",
       "26   0.811674  0.957355\n",
       "27   0.825369  0.986242\n",
       "28   0.851571  1.013064\n",
       "29   0.875812  1.038352\n",
       "..        ...       ...\n",
       "970  1.362286  0.487075\n",
       "971  1.364332  0.485861\n",
       "972  1.366952  0.484254\n",
       "973  1.370742  0.482999\n",
       "974  1.374617  0.482199\n",
       "975  1.379062  0.481410\n",
       "976  1.383927  0.480518\n",
       "977  1.386315  0.479590\n",
       "978  1.390629  0.478590\n",
       "979  1.395271  0.478092\n",
       "980  1.399911  0.477632\n",
       "981  1.403010  0.477391\n",
       "982  1.407117  0.477618\n",
       "983  1.410680  0.477404\n",
       "984  1.414096  0.477067\n",
       "985  1.416664  0.477249\n",
       "986  1.418140  0.477510\n",
       "987  1.419588  0.478258\n",
       "988  1.420723  0.478977\n",
       "989  1.421980  0.479671\n",
       "990  1.421662  0.480670\n",
       "991  1.420208  0.481684\n",
       "992  1.417377  0.482586\n",
       "993  1.416061  0.483551\n",
       "994  1.414663  0.484430\n",
       "995  1.414273  0.485187\n",
       "996  1.414419  0.485839\n",
       "997  1.414960  0.486591\n",
       "998  1.414727  0.487044\n",
       "999  1.413222  0.489630\n",
       "\n",
       "[1000 rows x 2 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt120_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
